README
NAME
    RTx::EmailCompletion - Add auto completion on RT email fields

VERSION
    This document describes version 0.02 of RTx::EmailCompletion.

DESCRIPTION
    I'm so tired to type email address by hand that I've done this module to
    add AJAX autocompletion on all email field of RT. As adding completion
    is dynamic, it should work on most RT releases (see later if it's not
    the case).

    There's 3 things :

    *   a small web service `html/Ajax/EmailCompletion' which search in all
        known users in the database ;

    *   prototype library to manipulate DOM and scriptaculous library to
        autocomplete ;

    *   a small javascript which parse html pages and add autocomplete on
        known input tags.

INSTALLATION
    if upgrading from 0.01, see later UPGRADE FROM 0.01.

    Install it like a standard perl module :

     RTHOME=/opt/rt3 perl Makefile.PL
     make
     make install

CONFIGURATION
    By default, completion works only for privileged users.

    You can activate it for unprivileged users (in the SelfService) by
    setting $EmailCompletionUnprivileged in RTHOME/etc/RT_SiteConfig.pm.
    There's three ways :

    * show everybody
          Set($EmailCompletionUnprivileged,"everybody");

    * show only privileged users
          Set($EmailCompletionUnprivileged,"privileged");

    * show only email matching a regexp
          Set($EmailCompletionUnprivileged, qr/\@my\.corp\.domain$/ );

HOW TO ADD FIELD TO AUTOCOMPLETION
    If you find email field without autocomplete, you can modify
    `html/NoAuth/js/emailcompletion.js' to handle this field (and email me
    to patch this module).

    At the beginning of this file you will find two global vars
    `multipleCompletion' and `singleCompletion'. They are array of regexp.

    Regexp must match all the word because `^' and `$' are added for
    matching. So if you want to match `Field1' and `Field2' you must add
    something like `Field.' or better `Field[12]'.

    To verify javascript find your input tag, you can uncomment the line
    just after the "DEBUGGING PURPOSE" one. All input tags find by the
    script will appear with a big red border.

UPGRADE FROM 0.01
    As directory structure has change, If you upgrade from 0.01, you must
    delete :

      RTHOME/local/html/Ajax/EmailCompletion
      RTHOME/local/html/NoAuth/js/emailcompletion.js
      RTHOME/local/html/NoAuth/js/

    Be careful of you have other javascripts in RTHOME/local/html/NoAuth/js/

HISTORY
    The first version (unreleased) modify html pages. The better method
    actually used allow this module to be compatible with, virtually, all RT
    release.

AUTHORS
    Nicolas Chuche <nchuche@barna.be>

COPYRIGHT
    Copyright 2007 by Nicolas Chuche <nchuche@barna.be>

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    See http://www.perl.com/perl/misc/Artistic.html

    All Scriptaculous and Prototype program are placed under MIT licence and
    are copyrighted by their owners (see top of files).



CHANGES
Changes for 0.02

  * allow email completion for unprivileged users (disable by default). See README
  * add debugging tools to find input tags found by emailcompletion.js

Changes for 0.01

  * first release

html/NoAuth/js/emailcompletion.js
function initPage() {
    addAutoComplete();
} // initPage

/* CUSTOMIZE PART */

// documentation can be found in lib/RTx/EmailCompletion.pm

// we need two Arrays :
// one for input with multiple email address allowed
var multipleCompletion = new Array("Requestors", "Cc", "AdminCc", "WatcherAddressEmail[123]", "UpdateCc", "UpdateBcc");

// and one for input with only one email address allowed
var singleCompletion   = new Array("(Add|Delete)Requestor", "(Add|Delete)Cc", "(Add|Delete)AdminCc");

/* END OF CUSTOMIZE PART */

var globalRegexp   = new RegExp('^(' + multipleCompletion.concat(singleCompletion).join('|') + ')$');
var multipleRegexp = new RegExp('^(' + multipleCompletion.join('|') + ')$');

function addAutoComplete() {
    var inputs = document.getElementsByTagName("input");
    
    for (var i = 0; i < inputs.length; i++) {
	var input = inputs[i];
	var inputName = input.getAttribute("name");

	// if empty or not defined
	if (! inputName)
	    continue;

	// only input's names defined in global vars at the beginning
	// are concerned
	if ( ! inputName.match(globalRegexp) )
	    continue;
	
	// if multiple email address are allowed we add the token
	// option to Autocompleter
	var options = '';
	if (inputName.match(multipleRegexp))
	    options = "tokens: ','";

	// we must set an id for scriptaculous
	// we use the name that seems to be uniq
	input.id = inputName;

	// DEBUGGING PURPOSE 
	// input.className += input.className? " emailcompletiondebug" : "emailcompletiondebug";

	var div = '<div class="autocomplete" id="' + inputName + '_to_auto_complete" />';
	div += '<script type="text/javascript">new Ajax.Autocompleter(\'' + inputName;
	div += "', '" + inputName + "_to_auto_complete', '<%$RT::WebPath%>/SelfService/Ajax/EmailCompletion\', {" + options + "})</script>";
	
	// use prototype to add the div after input
	new Insertion.After(inputName,div);
    }
	
} //addAutoComplete

Event.observe(window, 'load', initPage);


lib/RTx/EmailCompletion.pm
package RTx::EmailCompletion;

use strict;

our $VERSION = "0.02";

1;

__END__

=head1 NAME

RTx::EmailCompletion - Add auto completion on RT email fields

=head1 VERSION

This document describes version 0.02 of RTx::EmailCompletion.

=head1 DESCRIPTION

I'm so tired to type email address by hand that I've done this module
to add AJAX autocompletion on all email field of RT. As adding
completion is dynamic, it should work on most RT releases (see later
if it's not the case).

There's 3 S<things :>

=over

=item *

a small web service C<html/Ajax/EmailCompletion> which search in all
known users in the S<database ;>

=item *

prototype library to manipulate DOM and scriptaculous library to
S<autocomplete ;>

=item *

a small javascript which parse html pages and add autocomplete on
known input tags.

=back

=head1 INSTALLATION

if upgrading from 0.01, see later UPGRADE FROM 0.01.

Install it like a standard perl module :

 RTHOME=/opt/rt3 perl Makefile.PL
 make
 make install

=head1 CONFIGURATION

By default, completion works only for privileged users.

You can activate it for unprivileged users (in the SelfService) by
setting $EmailCompletionUnprivileged in
RTHOME/etc/RT_SiteConfig.pm. There's three ways :

=over

=item *
show everybody

  Set($EmailCompletionUnprivileged,"everybody");

=item *
show only privileged users

  Set($EmailCompletionUnprivileged,"privileged");

=item *
show only email matching a regexp

  Set($EmailCompletionUnprivileged, qr/\@my\.corp\.domain$/ );

=back

=head1 HOW TO ADD FIELD TO AUTOCOMPLETION

If you find email field without autocomplete, you can modify
C<html/NoAuth/js/emailcompletion.js> to handle this field (and email
me to patch this module).

At the beginning of this file you will find two global vars
C<multipleCompletion> and C<singleCompletion>. They are array of
regexp.

Regexp must match all the word because C<^> and C<$> are added for
matching. So if you want to match C<Field1> and C<Field2> you must add
something like C<Field.> or better C<Field[12]>.

To verify javascript find your input tag, you can uncomment the line
just after the "DEBUGGING PURPOSE" one. All input tags find by the
script will appear with a big red border.

=head1 UPGRADE FROM 0.01

As directory structure has change, If you upgrade from 0.01, you must
delete :

  RTHOME/local/html/Ajax/EmailCompletion
  RTHOME/local/html/NoAuth/js/emailcompletion.js
  RTHOME/local/html/NoAuth/js/

Be careful of you have other javascripts in RTHOME/local/html/NoAuth/js/

=head1 HISTORY

The first version (unreleased) modify html pages. The better method
actually used allow this module to be compatible with, virtually, all
RT release.

=head1 AUTHORS

Nicolas Chuche E<lt>nchuche@barna.beE<gt>

=head1 COPYRIGHT

Copyright 2007 by Nicolas Chuche E<lt>nchuche@barna.beE<gt>

This program is free software; you can redistribute it and/or 
modify it under the same terms as Perl itself.

See L<http://www.perl.com/perl/misc/Artistic.html>

All Scriptaculous and Prototype program are placed under MIT licence
and are copyrighted by their owners (see top of files).

=cut
